src/core/lib/surface/server.c

```diff
@@ -507,26 +521,37 @@ static void publish_new_rpc(grpc_exec_ctx *exec_ctx, void *arg, bool success) {
     return;
   }
 
-  int request_id = gpr_stack_lockfree_pop(rm->requests);
-  if (request_id == -1) {
-    gpr_mu_lock(&server->mu_call);
-    gpr_mu_lock(&calld->mu_state);
-    calld->state = PENDING;
-    gpr_mu_unlock(&calld->mu_state);
-    if (rm->pending_head == NULL) {
-      rm->pending_tail = rm->pending_head = calld;
+  for (size_t i = 0; i < server->cq_count; i++) {
+    size_t cq_idx = (chand->cq_idx + i) % server->cq_count;
+    int request_id = gpr_stack_lockfree_pop(rm->requests_per_cq[cq_idx]);
+    if (request_id == -1) {
+      continue;
     } else {
-      rm->pending_tail->pending_next = calld;
-      rm->pending_tail = calld;
+      gpr_log(GPR_DEBUG, "queue lockfree, retries=%d chose=%d", i, cq_idx);
```

Comment:
Oh... I should've noted: there will be a gpr_debug removal sweep prior to
this being done.

On Fri, May 20, 2016, 11:36 PM Sree Kuchibhotla <notifications@github.com>
wrote:

> In src/core/lib/surface/server.c
> <https://github.com/grpc/grpc/pull/6612#discussion_r64129786>:
>
> >      } else {
> > -      rm->pending_tail->pending_next = calld;
> > -      rm->pending_tail = calld;
> > +      gpr_log(GPR_DEBUG, "queue lockfree, retries=%d chose=%d", i, cq_idx);
>
> Delete the debug log here (and at lines 542, 1304, 1311, 1333, 1387 and
> 1402)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/6612/files/ae09d9dca9ac0f6d6c6e877e2935ad8cfba9da05#r64129786>
>